wrappers/internal/service_status.go

```diff
@@ -68,60 +78,168 @@ func (s *ServiceStatus) IsEnabled() bool {
+		Daemon:           unitStatus.Daemon,
+		Id:               unitStatus.Id,
+		Name:             unitStatus.Name,
+		Names:            unitStatus.Names,
+		Enabled:          unitStatus.Enabled,
+		Active:           unitStatus.Active,
+		Installed:        unitStatus.Installed,
+		NeedDaemonReload: unitStatus.NeedDaemonReload,
+	}
+}
+
+func findMatchingStatus(name string, services []client.ServiceUnitStatus) *systemd.UnitStatus {
+	for _, svc := range services {
+		if svc.Name == name {
+			return clientUnitStatusToSystemdUnitStatus(svc)
+		}
+	}
+	return nil
+}
+
+func queryUserServiceStatusMany(apps []*snap.AppInfo, units []string) (map[int][]*ServiceStatus, error) {
+	usrUnitStss, _, err := userSessionQueryServiceStatusMany(units)
 	if err != nil {
 		return nil, err
 	}
 
-	var appStatuses []*ServiceStatus
-	var statusIndex int
+	constructStatus := func(app *snap.AppInfo, uid int) *ServiceStatus {
+		primarySvcName, activators := SnapServiceUnits(app)
+		primarySvcUnit := findMatchingStatus(primarySvcName, usrUnitStss[uid])
+		if primarySvcUnit == nil {
+			return nil
+		}
+
+		svcSt := &ServiceStatus{
+			name:        app.Name,
+			user:        true,
+			service:     primarySvcUnit,
+			slotEnabled: serviceIsSlotActivated(app),
+		}
+		if len(activators) > 0 {
+			for _, act := range activators {
+				actSvcUnit := findMatchingStatus(act, usrUnitStss[uid])
+				if actSvcUnit == nil {
+					return nil
+				}
+				svcSt.activators = append(svcSt.activators, actSvcUnit)
+			}
+		}
+		return svcSt
+	}
+
+	// For each user we have results from, go through services and build a list of service results
+	svcsStatusMap := make(map[int][]*ServiceStatus)
+	for uid := range usrUnitStss {
+		var svcs []*ServiceStatus
+		for _, app := range apps {
+			if !app.IsService() {
+				continue
+			}
+			if app.DaemonScope != snap.UserDaemon {
+				continue
+			}
+			svc := constructStatus(app, uid)
+			if svc == nil {
+				// In theory should not happen, we either receive *all* requested statuses from the REST service
+				// or none if something goes wrong with querying one of them. If we receive none, then the entry
+				// won't exist in usrUnitStss and we shouldn't even be in this loop.
+				return nil, fmt.Errorf("internal error: no status received for service %s", app.ServiceName())
+			}
+			svcs = append(svcs, svc)
+		}
+		svcsStatusMap[uid] = svcs
+	}
+	return svcsStatusMap, nil
+}
+
+func querySystemServiceStatusMany(sysd systemd.Systemd, apps []*snap.AppInfo, units []string) ([]*ServiceStatus, error) {
+	sysUnitStss, err := sysd.Status(units)
+	if err != nil {
+		return nil, err
+	}
+
+	var sysIndex int
+	getStatus := func(app *snap.AppInfo, activators []string) *ServiceStatus {
+		svcSt := &ServiceStatus{
+			name:        app.Name,
+			service:     sysUnitStss[sysIndex],
+			slotEnabled: serviceIsSlotActivated(app),
+		}
+		if len(activators) > 0 {
+			svcSt.activators = sysUnitStss[sysIndex+1 : sysIndex+1+len(activators)]
```

Comment:
the activators case is not reached by unit tests

Reply:
added tests for this, thank you